### ISSUE TITLE
why not using size (n_pre, n_cur) as the input to the Affinity Net?

### ISSUE THREAD

**Issue by nejyeah** [shijieS/SST#3]
Hi, 

Thanks for sharing the code. 
In your code, the output features of selector are with shape (80, 520) and the inputs to affinity net are with shape(1040, 80, 80).
Why not set the output features of selector as (n_obj, 520), and then the inputs to affinity net will be with shape (1040, n_pre, n_cur). It seems okay with variable input size to FCN.  And it will save time for training and inference.

**Comment by shijieS** [comment]
Thank you for your suggestions.  It will be great to output features of selector as (n_obj, 520). It does save time for training. However, as far as I know, the input size of FCN cannot get changed. If not, could you provide me with some examples?

**Comment by nejyeah** [comment]
Thanks for your reply. Actually I am not that familiar with Pytorch.  I will try and report the results.

**Comment by shijieS** [comment]
Look forward to your good results :) .

**Comment by SpyderXu** [comment]
> Thank you for your suggestions. It will be great to output features of selector as (n_obj, 520). It does save time for training. However, as far as I know, the input size of FCN cannot get changed. If not, could you provide me with some examples?

In semantic segmentation, it seems that the FCN network could deal with any size input image.

**Comment by shijieS** [comment]
> > Thank you for your suggestions. It will be great to output features of selector as (n_obj, 520). It does save time for training. However, as far as I know, the input size of FCN cannot get changed. If not, could you provide me with some examples?
> 
> In semantic segmentation, it seems that the FCN network could deal with any size input image.

Thank you for your reply, @SpyderXu. The FCN can be referred to [torch.nn.functional.linear](https://pytorch.org/docs/stable/nn.html#torch.nn.functional.linear).  The *in_features* is fixed. Unfortunately, I am not familar with semantic segmentation. If possible, could you provide some more details?

**Comment by shijieS** [comment]
Thank you for your reply, @SpyderXu . Now, I understand that. @nejyeah , Your idea is really great. I think this network can be greatly improved.

**Comment by nejyeah** [comment]
Hi, the problem of varying size seems to locate at the trainning batch. In a Batch, we need to make the features to have the same size. The training loss may be modified to discard the padding value when using a fixed input size. And then the inference can take advantage of the varying size for just one frame each time.

**Comment by nejyeah** [comment]
In the training process, the row and col of the labels have been shuffled so as the input feature(1040*80*80) before feeding into the compression network. Because the compression network just consists of 1x1 conv, the shuffle of the order(80*80) seems to be useless.  

I am a beginner to the tracking. So I am grateful to you for sharing the nice project. I have learned a lot from it.

**Comment by nejyeah** [comment]
> Actually, the feature extraction process can be put in the detector (such as SSD), which will also improve this framework. I need to do that.

Good idea! For practical, the detect net and MOT net should be merged into one.

**Comment by shijieS** [comment]
> In the training process, the row and col of the labels have been shuffled so as the input feature(1040 x 80 x 80) before feeding into the compression network. Because the compression network just consists of 1x1 conv, the shuffle of the order(80 x 80) seems to be useless.

Yes. as you say, the shuffle of the order is useless.  @nejyeah . Thank you for your appreciation of my work.  It's also my first work on tracking. Thank you for your suggestions again.

**Comment by shijieS** [comment]
> Hi, the problem of varying size seems to locate at the trainning batch. In a Batch, we need to make the features to have the same size. The training loss may be modified to discard the padding value when using a fixed input size. And then the inference can take advantage of the varying size for just one frame each time.

I add dummy box into the *MOTTrainDataset*. So you get modified from here and it should be lots of work.  The training loss uses mask in order to remove the negative affect of dummy box. All the work is based on my first thinking which is not that perfect.

**Comment by nejyeah** [comment]
I have tested the varying size for inference on GTX1080Ti for MOT17-01-FRCNN. For the original version,  the inference time is about 0.266 s/frame. For the varying size,  the inference time is about 0.243.  And It saves 20 ms.  And the results seems to make little difference from eyes based on the output vedio.

**Comment by shijieS** [comment]
That's great. By the way, different input boxes may take different time. Is this the average time. If possible, could you push a branch to this repository?

**Comment by nejyeah** [comment]
Yes, it is the average time in your code for tracking test.
I just modified two functions in your code.
Modify layers/sst.py
```
126     def forward_stacker_features(self, xp, xn, fill_up_column=True):
127         pre_rest_num = self.max_object - xp.shape[1]
128         next_rest_num = self.max_object - xn.shape[1]
129         pre_num = xp.shape[1]
130         next_num = xn.shape[1]
131         ### x = self.forward_stacker2(
132         ###     self.resize_dim(xp, pre_rest_num, dim=1),
133         ###     self.resize_dim(xn, next_rest_num, dim=1)
134         ### )
            x = self.forward_stacker2_inference(xp, xn)
135       
136         x = self.final_dp(x)
137         # [B, N, N, 1]
138         x = self.forward_final(x, self.final_net)
139         x = x.contiguous()
140         # add zero
141         ### if next_num < self.max_object:
142         ###     x[0, 0, :, next_num:] = 0
143         ### if pre_num < self.max_object:
144         ###     x[0, 0, pre_num:, :] = 0
145         x = x[0, 0, :]
146         # add false unmatched row and column
147         x = self.resize_dim(x, 1, dim=0, constant=self.false_constant)
148         x = self.resize_dim(x, 1, dim=1, constant=self.false_constant)
149 
150         x_f = F.softmax(x, dim=1)
151         x_t = F.softmax(x, dim=0)
152         # slice
153         last_row, last_col = x_f.shape
154         row_slice = list(range(pre_num)) + [last_row-1]
155         col_slice = list(range(next_num)) + [last_col-1]
156         ### x_f = x_f[row_slice, :]
157         ### x_f = x_f[:, col_slice]
158         ### x_t = x_t[row_slice, :]
159         ### x_t = x_t[:, col_slice]
```
Add function forward_stacker2_inference(), copy and revised from forward_statcker2 for not affecting training 
```
223     def forward_stacker2_inference(self, stacker1_pre_output, stacker1_next_output):
224         ### stacker1_pre_output = stacker1_pre_output.unsqueeze(2).repeat(1, 1, self.max_object, 1).permute(0, 3, 1, 2)
225         ### stacker1_next_output = stacker1_next_output.unsqueeze(1).repeat(1, self.max_object, 1, 1).permute(0, 3, 1, 2)
            n_pre = stacker1_pre_output.shape[1]
            n_next = stacker1_next_output.shape[1]
            stacker1_pre_output = stacker1_pre_output.unsqueeze(2).repeat(1, 1, n_next, 1).permute(0, 3, 1, 2)
            stacker1_next_output = stacker1_next_output.unsqueeze(1).repeat(1, n_pre, 1, 1).permute(0, 3, 1, 2)
226 
227         stacker1_pre_output = self.stacker2_bn(stacker1_pre_output.contiguous()    )
228         stacker1_next_output = self.stacker2_bn(stacker1_next_output.contiguous    ())
229 
230         output = torch.cat(
231             [stacker1_pre_output, stacker1_next_output],
232             1
233         )
234 
235         return output
```

**Comment by shijieS** [comment]
Great, Thank you.

**Comment by yaoqi-zd** [comment]
@nejyeah @shijieS  Hi, I find that when the feature recorder compute the similarity between current frame and previous frames, it takes near 60ms for the first computation.
the time evaluation code fragment in "tracker.py" is like follows:
```
self.all_similarity[frame_index] = {}
            print("len(frames) = {}".format(len(self.all_frame_index)))
            for pre_index in self.all_frame_index[:-1]:
                delta = pow(TrackerConfig.decay, (frame_index - pre_index)/3.0)
                start = time.time()
                pre_similarity = sst.forward_stacker_features(self.all_features[pre_index], features, fill_up_column=False)
                self.all_similarity[frame_index][pre_index] = pre_similarity*delta
                print("stacker: {}".format(time.time() - start))
```
 and the results show:
```
len(frames) = 14                                                                                      │| NVIDIA-SMI 384.145                Driver Version: 384.145                   |
stacker: 0.05677461624145508                                                                          │|-------------------------------+----------------------+----------------------+
stacker: 0.0012462139129638672                                                                        │| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
stacker: 0.0012083053588867188                                                                        │| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
stacker: 0.0012061595916748047                                                                        │|===============================+======================+======================|
stacker: 0.0012083053588867188                                                                        │|   0  Tesla P100-PCIE...  On   | 00000000:05:00.0 Off |                    0 |
stacker: 0.001203298568725586                                                                         │| N/A   29C    P0    24W / 250W |      0MiB / 16276MiB |      0%      Default |
stacker: 0.0011980533599853516                                                                        │+-------------------------------+----------------------+----------------------+
stacker: 0.0012035369873046875                                                                        │|   1  Tesla P100-PCIE...  On   | 00000000:87:00.0 Off |                    0 |
stacker: 0.0011982917785644531                                                                        │| N/A   28C    P0    26W / 250W |      0MiB / 16276MiB |      0%      Default |
stacker: 0.0012028217315673828                                                                        │+-------------------------------+----------------------+----------------------+
stacker: 0.001201629638671875                                                                         │
stacker: 0.0011990070343017578                                                                        │+-----------------------------------------------------------------------------+
stacker: 0.0012001991271972656
```
do you have any ideas about this? I think this is the bottleneck of FPS for SST tracker
